kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,700 @@
+			}
+		}
+		log.Debugf("Found %d Service LoadBalancer in Calico datastore", len(filteredServices))
+		return filteredServices, nil
+	}
+
+	cacheArgs := rcache.ResourceCacheArgs{
+		ListFunc:    listFunc,
+		ObjectType:  reflect.TypeOf(&v1.Service{}),
+		LogTypeDesc: "Service",
+	}
+	ccache := rcache.NewResourceCache(cacheArgs)
+
+	svcHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			log.Debugf("Got ADD event for service")
+			svc := obj.(*v1.Service)
+			if isCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
+				ccache.Set(createHandle(svc), svc)
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			log.Debugf("Got UPDATE event for service")
+			log.Debugf("Old object: %v", oldObj)
+			log.Debugf("New object: %v", newObj)
+			svcNew := newObj.(*v1.Service)
+			svcOld := oldObj.(*v1.Service)
+
+			// Service changed type, we need to release the addresses used by the service
+			if svcNew.Spec.Type != v1.ServiceTypeLoadBalancer &&
+				svcOld.Spec.Type == v1.ServiceTypeLoadBalancer {
+				ccache.Delete(createHandle(svcOld))
+			} else if svcOld.Annotations[annotationIpv4Pools] != svcNew.Annotations[annotationIpv4Pools] ||
+				svcOld.Annotations[annotationIpv6Pools] != svcNew.Annotations[annotationIpv6Pools] ||
+				svcOld.Annotations[annotationLoadBalancerIp] != svcNew.Annotations[annotationLoadBalancerIp] {
+				// Calico annotations have changed, get new address based on new conditions.
+				ccache.Set(createHandle(svcNew), svcNew)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			log.Debugf("Got DELETE event for service")
+			svc := obj.(*v1.Service)
+			if isCalicoManagedLoadBalancer(svc, cfg.AssignIPs) {
+				ccache.Delete(createHandle(svc))
+			}
+		},
+	}
+	_, informer := cache.NewIndexerInformer(svcWatcher, &v1.Service{}, 0, svcHandler, cache.Indexers{})
+
+	return &loadBalancerController{
+		calicoClient:    calicoClient,
+		ctx:             ctx,
+		cfg:             cfg,
+		clientSet:       clientset,
+		dataFeed:        node.NewDataFeed(calicoClient),
+		lbResourceCache: ccache,
+		informer:        informer,
+		syncerUpdates:   make(chan interface{}),
+		syncChan:        make(chan interface{}),
+		ipamBlocks:      make(map[string]model.KVPair),
+		ipPools:         make(map[string]api.IPPool),
+	}
+}
+
+// Run starts the controller.
+func (c *loadBalancerController) Run(stopCh chan struct{}) {
+	defer uruntime.HandleCrash()
+
+	// Start the Kubernetes informer, which will start syncing with the Kubernetes API.
+	log.Info("Starting LoadBalancer controller")
+	go c.informer.Run(stopCh)
+
+	// Wait until we are in sync with the Kubernetes API before starting the
+	// resource cache.
+	log.Debug("Waiting to sync with Kubernetes API (Service)")
+	for !c.informer.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+	log.Debug("Finished syncing with Kubernetes API (Service)")
+	c.lbResourceCache.Run(timer.String())
+
+	for i := 0; i < c.cfg.NumberOfWorkers; i++ {
```

Comment:
We don't need multiple workers in this controller - in fact, doing the logic of this controller in parallel will break things like IPAM GC because each worker will be unaware of the actions the other workers are taking, so we should just get rid of this.